core/src/main/java/org/axonframework/eventsourcing/eventstore/inmemory/InMemoryEventStorageEngine.java

```diff
@@ -93,6 +95,27 @@ public DomainEventStream readEvents(String aggregateIdentifier, long firstSequen
         return Optional.ofNullable(snapshots.get(aggregateIdentifier));
     }
 
+    @Override
+    public TrackingToken createHeadToken() {
+        if (events.size() == 0) {
+            return null;
+        }
+        return events.lastKey();
+    }
+
+    @Override
+    public TrackingToken createTokenAt(Instant dateTime) {
+        return events.values()
+                     .stream()
+                     .filter(event -> event.getTimestamp().equals(dateTime) || event.getTimestamp().isAfter(dateTime))
+                     .min(Comparator.comparingLong(e -> ((GlobalSequenceTrackingToken) e.trackingToken())
```

Comment:
I assume you're sure you can make this `GlobalSequenceTrackingToken` cast?
If so, you could add a `map` to perform the cast.

Reply:
Yes, `InMemoryEventStorageEngine` works only with `GlobalSequenceToken`s.